### DISCUSSION TITLE
Using ContextVar for scoped session

### DISCUSSION THREAD

**Discussion by manusinghal19** [sqlalchemy/sqlalchemy#9035] [Ideas]
I was going through the source code of scoped_session, and since we're going to support python3.7+ only from the 2.0 version onwards, Shouldn't we use ContextVar to manage scoped sessions?

If yes, I can take it up. If not, would like to know the possible concerns.

Thanks in Advance

**Comment by zzzeek** [comment]
hey there -

note we do more kinds of context than just threading.local().   It looks like ContextVar allows other kinds of context, sort of, but im not really following the docs very well just looking.

what's the benefit we are going for?  Note that 2.0 is about to release 2.0.0rc1 and we'd like to be done making big changes, final release is for January sometime.   So I'd rather not add more architectural changes without a really good reason.

**Reply by manusinghal19** [reply]
Thanks for quick response @zzzeek . 
1. This is definitely not something I am suggesting for 2.0 release, but post that in some major release, that too if it makes sense.

2. To my understanding, contextVar can handle two scenarios - multithreading and asyncio. and these are the only 2 scenarios I think one may need scoped sessions. If there are more scenarios, SQLAlchemy's current approach is way more flexible as it allows users to use anything which meets their need as scopefunc.

3. Having said that, the only reason why contextVar should be used is because, if language itself is giving support for some common problem, that's a preference and better code readability. plus, maybe, there could be some performance benefits as well.

Looking forward for more clarity on this.

Thanks

**Reply by zzzeek** [reply]
the ScopedSession can scope to other things like HTTP request objects so in that sense it's more generalized than ContextVar.

the [ScopedRegistry](https://github.com/sqlalchemy/sqlalchemy/blob/f74de7bc0b886b3585536e2a3c7f1828ec8be083/lib/sqlalchemy/util/_collections.py#L600) and [ThreadLocalRegistry](https://github.com/sqlalchemy/sqlalchemy/blob/f74de7bc0b886b3585536e2a3c7f1828ec8be083/lib/sqlalchemy/util/_collections.py#L664) have been mostly unchanged for over a decade and IMO they are not hard to read.   It seems a bit unnecessary and risk-introducing to change their implementation to one that does not provide as much extensibility and also has been known to [cause](https://github.com/python-greenlet/greenlet/issues/196)  [issues](https://github.com/gevent/gevent/issues/1407) .

**Reply by manusinghal19** [reply]
Thanks. This information is very useful.

**Reply by zzzeek** [reply]
you know what might be useful is to have the option to use a ContextVar as the scopefunc.   We can add that pretty easily, how about that?  so if an app wants to use ContextVars for things, as they should, they can have their ScopedSession use that ContextVar also.  how about that?

**Reply by CaselIT** [reply]
> you know what might be useful is to have the option to use a ContextVar as the scopefunc.

I wanted to suggest it yesterday but than forgot

**Reply by manusinghal19** [reply]
Yes, That's a good idea. Thanks.
